crates/bevy_ecs/src/system/system_registry.rs

```diff
@@ -0,0 +1,277 @@
+/// #[derive(Resource, Default)]
+/// struct ChangeDetector;
+///
+/// let mut world = World::default();
+/// world.init_resource::<ChangeDetector>();
+/// let detector = world.register_system(|change_detector: ResMut<ChangeDetector>| {
+///     if change_detector.is_changed() {
+///         println!("Something happened!");
+///     } else {
+///         println!("Nothing happaned.");
+///     }
+/// });
+///
+/// // Resources are changed when they are first added
+/// let _ = world.run_system_by_id(detector); // -> Something happened!
+/// let _ = world.run_system_by_id(detector); // -> Nothing happened.
+/// world.resource_mut::<ChangeDetector>().set_changed();
+/// let _ = world.run_system_by_id(detector); // -> Something happened!
+/// ```
+
+#[derive(Resource, Default)]
+pub struct SystemRegistry {
+    last_id: u32,
+    systems: HashMap<u32, (bool, BoxedSystem)>,
+}
+
+/// An identifier for a system registered in the [`SystemRegistry`].
+#[derive(Debug, Clone, Copy)]
+pub struct SystemId(u32);
+
+impl SystemRegistry {
+    /// Registers a system in the [`SystemRegistry`], so it can be run later.
+    ///
+    /// It's possible to register the same system twice, creating two instances in the registry.
+    #[inline]
+    pub fn register<M, S: IntoSystem<(), (), M> + 'static>(&mut self, system: S) -> SystemId {
+        let id = self.last_id + 1;
+        self.last_id = id;
+        self.systems
+            .insert(id, (false, Box::new(IntoSystem::into_system(system))));
+        SystemId(id)
+    }
+
+    /// Removes a registered system from the [`SystemRegistry`], if the [`SystemId`] is not
+    /// registered, this function does nothing.
+    #[inline]
+    pub fn remove(&mut self, id: SystemId) {
+        self.systems.remove(&id.0);
+    }
+
+    /// Run the system by its [`SystemId`].
+    ///
+    /// Systems must be registered before they can be run.
+    #[inline]
+    pub fn run_by_id(
+        &mut self,
+        world: &mut World,
+        id: SystemId,
+    ) -> Result<(), SystemRegistryError> {
+        match self.systems.get_mut(&id.0) {
+            Some((initialized, matching_system)) => {
+                if !*initialized {
+                    matching_system.initialize(world);
+                    *initialized = true;
+                }
+                matching_system.run((), world);
+                matching_system.apply_deferred(world);
+                Ok(())
+            }
+            None => Err(SystemRegistryError::SystemIdNotRegistered(id)),
+        }
+    }
+}
+
+impl World {
+    /// Registers a system in the [`SystemRegistry`].
+    ///
+    /// Calls [`SystemRegistry::register`].
+    #[inline]
+    pub fn register_system<M, S: IntoSystem<(), (), M> + 'static>(
+        &mut self,
+        system: S,
+    ) -> SystemId {
+        if !self.contains_resource::<SystemRegistry>() {
+            panic!(
+                "SystemRegistry not found: Nested and recursive one-shot systems are not supported"
```

Comment:
These warnings don't match what you said about recursive one-shot systems in the PR thread :)

Reply:
Maybe there has been a misunderstanding.
Recursive / nested systems do not work with `run_system_by_id`, while `run_system` can be recursed and nested.